From 4bc69a18c99bc6431ceca3fd681d303970e7854c Mon Sep 17 00:00:00 2001
From: Thomas Bourgoin <thomas.bourgoin@foss.st.com>
Date: Mon, 17 Feb 2025 18:29:13 +0100
Subject: [PATCH 795/954] refactor(st-bsec): fix bsec state initialisation

The value returned by bsec_get_state() give the status of BSEC.
BSEC peripheral should always be closed.
BSEC open is not supported by OpenSTLinux.

Remove function bsec_mode_is_closed_device() as it does the same as
stm32_otp_is_closed_device().

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 598f082dad3b50e96273ed7813aae588a29063fe refactor(st-bsec): fix bsec state initialisation]
Signed-off-by: Thomas Bourgoin <thomas.bourgoin@foss.st.com>
Change-Id: If391bb78f1868955e170e635ace397c01be4e0d6
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/437070
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
---
 drivers/st/bsec/bsec2.c                     | 47 +++------------------
 drivers/st/bsec/bsec3.c                     | 18 ++++----
 include/drivers/st/bsec.h                   | 21 ++++-----
 plat/st/stm32mp2/include/stm32mp2_private.h |  4 +-
 4 files changed, 26 insertions(+), 64 deletions(-)

diff --git a/drivers/st/bsec/bsec2.c b/drivers/st/bsec/bsec2.c
index 14df63c134..70bcfee53f 100644
--- a/drivers/st/bsec/bsec2.c
+++ b/drivers/st/bsec/bsec2.c
@@ -822,52 +822,17 @@ uint32_t bsec_check_nsec_access_rights(uint32_t otp)
 }
 #endif
 
-uint32_t bsec_get_secure_state(void)
+uint32_t bsec_get_state(void)
 {
 	uint32_t status = bsec_get_status();
-	uint32_t result = BSEC_STATE_INVALID;
-	uint32_t otp_enc_id __unused;
-	uint32_t otp_bit_len __unused;
-	int res __unused;
 
-	if ((status & BSEC_OTP_STATUS_INVALID) != 0U) {
-		result = BSEC_STATE_INVALID;
-	} else {
-		if ((status & BSEC_OTP_STATUS_SECURE) != 0U) {
-			if (stm32mp_check_closed_device() == STM32MP_CHIP_SEC_CLOSED) {
-				result = BSEC_STATE_SEC_CLOSED;
-			} else {
-				result = BSEC_STATE_SEC_OPEN;
-			}
-		} else {
-			/* OTP modes OPEN1 and OPEN2 are not supported */
-			result = BSEC_STATE_INVALID;
-		}
+	if ((status & BSEC_OTP_STATUS_INVALID) == BSEC_OTP_STATUS_INVALID) {
+		return BSEC_STATE_INVALID;
 	}
 
-#if STM32MP13
-	res = stm32_get_otp_index(ENCKEY_OTP, &otp_enc_id, &otp_bit_len);
-	if ((res == 0) && (otp_bit_len > 0)) {
-		uint32_t start = otp_enc_id / sizeof(uint32_t);
-		size_t otp_nb = round_up(otp_bit_len, sizeof(uint32_t)) /
-				sizeof(uint32_t);
-		size_t otp_locked_nb = 0U;
-		uint32_t idx;
-
-		for (idx = start; idx < start + otp_nb; idx++) {
-			bool locked = false;
-
-			res = bsec_read_sp_lock(idx, &locked);
-			if ((res != 0U) || (locked == 0U)) {
-				break;
-			}
-			otp_locked_nb++;
-		}
-		if (otp_locked_nb == otp_nb) {
-			result |= BSEC_HARDWARE_KEY;
-		}
+	if ((status & BSEC_OTP_STATUS_SECURE) == BSEC_OTP_STATUS_SECURE) {
+		return BSEC_STATE_CLOSED;
 	}
-#endif
 
-	return result;
+	return BSEC_STATE_INVALID;
 }
diff --git a/drivers/st/bsec/bsec3.c b/drivers/st/bsec/bsec3.c
index cb42b870a9..383093b37c 100644
--- a/drivers/st/bsec/bsec3.c
+++ b/drivers/st/bsec/bsec3.c
@@ -458,10 +458,10 @@ uint32_t bsec_read_sp_lock(uint32_t otp, bool *value)
 }
 
 /*
- * bsec_get_secure_state: read state in BSEC status register.
- * return: secure state
+ * bsec_get_state: read state in BSEC status register.
+ * return: bsec state
  */
-uint32_t bsec_get_secure_state(void)
+uint32_t bsec_get_state(void)
 {
 	uint32_t state = BSEC_STATE_INVALID;
 	uint32_t status = bsec_get_status();
@@ -469,12 +469,14 @@ uint32_t bsec_get_secure_state(void)
 
 	if ((status & BSEC_OTPSR_INIT_DONE) == BSEC_OTPSR_INIT_DONE) {
 		/* NVSTATE is only valid if INIT_DONE */
-		uint32_t nvstates = (bsec_sr & BSEC_SR_NVSTATE_MASK) >> BSEC_SR_NVSTATE_SHIFT;
+		uint32_t nvstate = (bsec_sr & BSEC_SR_NVSTATE_MASK) >>
+				   BSEC_SR_NVSTATE_SHIFT;
 
-		if (nvstates == BSEC_SR_NVSTATE_OPEN)
-			state = BSEC_STATE_SEC_OPEN;
-		else if (nvstates == BSEC_SR_NVSTATE_CLOSED)
-			state = BSEC_STATE_SEC_CLOSED;
+		if (nvstate != BSEC_SR_NVSTATE_CLOSED) {
+			ERROR("BSEC invalid nvstate %#x\n", nvstate);
+		} else {
+			state = BSEC_STATE_CLOSED;
+		}
 	}
 
 	if ((bsec_sr & BSEC_SR_HVALID) != 0U) {
diff --git a/include/drivers/st/bsec.h b/include/drivers/st/bsec.h
index ac1685def3..00af06bdcc 100644
--- a/include/drivers/st/bsec.h
+++ b/include/drivers/st/bsec.h
@@ -27,18 +27,17 @@
 #define BSEC_WRITE_LOCKED		0xFFFFFFF6U
 
 /*
- * get BSEC global state: result for bsec_get_secure_state()
+ * get BSEC global state: result for bsec_get_state()
+ * BSEC open is considered invalid in OSTL use case.
  * @state: global state
  *           [1:0] BSEC state
- *             00b: Sec Open
- *             01b: Sec Closed
- *             11b: Invalid
+ *             0b: Closed
+ *             1b: Invalid
  *           [8]: Hardware Key set = 1b
  */
-#define BSEC_STATE_SEC_OPEN		U(0x0)
-#define BSEC_STATE_SEC_CLOSED		U(0x1)
-#define BSEC_STATE_INVALID		U(0x3)
-#define BSEC_STATE_MASK			GENMASK_32(1, 0)
+#define BSEC_STATE_CLOSED		U(0x0)
+#define BSEC_STATE_INVALID		U(0x1)
+#define BSEC_STATE_MASK			BIT(0)
 
 #define BSEC_HARDWARE_KEY		BIT(8)
 
@@ -61,11 +60,7 @@ uint32_t bsec_read_sw_lock(uint32_t otp, bool *value);
 uint32_t bsec_set_sp_lock(uint32_t otp);
 uint32_t bsec_read_sp_lock(uint32_t otp, bool *value);
 
-uint32_t bsec_get_secure_state(void);
-static inline bool bsec_mode_is_closed_device(void)
-{
-	return (bsec_get_secure_state() & BSEC_STATE_MASK) == BSEC_STATE_SEC_CLOSED;
-}
+uint32_t bsec_get_state(void);
 
 #if defined(IMAGE_BL32) || defined(STM32MP_SSP)
 uint32_t bsec_permanent_lock_otp(uint32_t otp);
diff --git a/plat/st/stm32mp2/include/stm32mp2_private.h b/plat/st/stm32mp2/include/stm32mp2_private.h
index ecd0c585f2..a4392af57c 100644
--- a/plat/st/stm32mp2/include/stm32mp2_private.h
+++ b/plat/st/stm32mp2/include/stm32mp2_private.h
@@ -91,12 +91,12 @@ static inline uint32_t stm32_otp_read_sw_lock(uint32_t otp, bool *value)
 
 static inline bool stm32_otp_is_closed_device(void)
 {
-	return bsec_mode_is_closed_device();
+	return (bsec_get_state() & BSEC_STATE_MASK) == BSEC_STATE_CLOSED;
 }
 
 static inline bool stm32_otp_is_hwkey_valid(void)
 {
-	if ((bsec_get_secure_state() & BSEC_HARDWARE_KEY) != 0U) {
+	if ((bsec_get_state() & BSEC_HARDWARE_KEY) != 0U) {
 		return true;
 	}
 
-- 
2.53.0

